opentracing-examples/src/test/java/io/opentracing/examples/actor_propagation/ActorPropagationTest.java

```diff
@@ -0,0 +1,125 @@
+
+import io.opentracing.ActiveSpan;
+import io.opentracing.mock.MockSpan;
+import io.opentracing.mock.MockTracer;
+import io.opentracing.mock.MockTracer.Propagator;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.ThreadLocalActiveSpanSource;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.Phaser;
+
+import static io.opentracing.examples.TestUtils.getOneByTag;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * These tests are intended to simulate the kind of async models that are common in java async
+ * frameworks.
+ *
+ * For improved readability, ignore the phaser lines as those are there to ensure deterministic
+ * execution for the tests without sleeps.
+ *
+ * @author tylerbenson
+ */
+public class ActorPropagationTest {
+
+  private final MockTracer tracer =
+      new MockTracer(new ThreadLocalActiveSpanSource(), Propagator.TEXT_MAP);
+  private Phaser phaser;
+
+  @Before
+  public void before() {
+    phaser = new Phaser();
+  }
+
+  @Test
+  public void testActorTell() {
+    try (Actor actor = new Actor(tracer, phaser)) {
+      phaser.register();
+      try (ActiveSpan parent =
+          tracer
+              .buildSpan("actorTell")
+              .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_PRODUCER)
+              .withTag(Tags.COMPONENT.getKey(), "example-actor")
+              .startActive()) {
+        actor.tell("my message");
+      }
+      phaser.arriveAndAwaitAdvance(); // child tracer started
+      assertThat(tracer.finishedSpans().size()).isEqualTo(0);
+      phaser.arriveAndAwaitAdvance(); // continue...
+
+      phaser.arriveAndAwaitAdvance(); // child tracer finished
+      assertThat(tracer.finishedSpans().size()).isEqualTo(1);
+      assertThat(getOneByTag(tracer.finishedSpans(), Tags.SPAN_KIND, Tags.SPAN_KIND_CONSUMER))
+          .isNotNull();
+      phaser.arriveAndAwaitAdvance(); // continue...
+
+      phaser.arriveAndAwaitAdvance(); // parent tracer finished
+      List<MockSpan> finished = tracer.finishedSpans();
+      phaser.arriveAndDeregister(); // continue...
+
+      assertThat(finished.size()).isEqualTo(2);
+      assertThat(finished.get(0).context().traceId())
+          .isEqualTo(finished.get(1).context().traceId());
+      assertThat(getOneByTag(finished, Tags.SPAN_KIND, Tags.SPAN_KIND_CONSUMER)).isNotNull();
+      assertThat(getOneByTag(finished, Tags.SPAN_KIND, Tags.SPAN_KIND_PRODUCER)).isNotNull();
+      assertThat(tracer.activeSpan()).isNull();
+    }
+  }
+
+  @Test
+  public void testActorAsk() throws ExecutionException, InterruptedException {
+    try (Actor actor = new Actor(tracer, phaser)) {
+      phaser.register();
+      Future<String> future;
+      try (ActiveSpan parent =
+          tracer
+              .buildSpan("actorAsk")
+              .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_PRODUCER)
```

Comment:
This should be `SPAN_KIND_CONSUMER`?

Reply:
Are you sure?  Isn't this where the messages are being generated and "sent" to the actor, which is doing the consuming?

Reply:
Yes, ignore.